### PR TITLE
basic setup for receiving mails and clicking on them

### PR DESCRIPTION
**Pull request by fatboypunk**

### CODE REVIEW

File: test/support/links/_post_show.html [hunk 1]
```diff
@@ -30,6 +34,7 @@ <h1>Post</h1>
                test-value="1"
                test-selector="post-123-upvote-comment" >
               Upvote Comment
+            </a>
```
Comment (EasterPeanut) [inline test/support/links/_post_show.html:37]: 👀

File: lib/test_dispatch.ex [hunk 2]
```diff
@@ -205,4 +205,55 @@ defmodule TestDispatch do
...
+      iex> if result =~ "Posts 123", do: :ok
+  """
+  def receive_mail(conn, %{subject: subject}) do
+    email =
+      receive do
+        {:delivered_email, %{subject: ^subject} = email} ->
+          email
+      after
+        100 ->
+          raise("Failed to receive any email")
```
Comment (EasterPeanut) [inline lib/test_dispatch.ex:231]:
```suggestion
```

File: lib/test_dispatch.ex [hunk 3]
```diff
@@ -205,4 +205,55 @@ defmodule TestDispatch do
...
+
```
Comment (EasterPeanut) [inline lib/test_dispatch.ex:232]:
```suggestion
```

File: README.md [hunk 4]
```diff
@@ -4,9 +4,9 @@
 
 # TestDispatch
 
-Helper to test the dispatch of Phoenix forms in Elixir applications. This will
-make it easier to write integration tests to check if forms in Phoenix templates
-will submit to the intended controller action with the right params.
+A library that adds the ability to use controller tests as integration tests
```
Comment (fatboypunk) [inline README.md:7]:
```suggestion
TestDispatch adds the ability to use controller tests as integration tests
```

File: README.md [hunk 5]
```diff
@@ -90,3 +92,32 @@ this will result in the following params:
 Ultimately, the conn is dispatched to the conn's `private.phoenix_endpoint`
 using `Phoenix.ConnTest.dispatch/5`, with the params and with the method and
 action found in the form.
+
+## Clicking on links in mails
```
Comment (fatboypunk) [inline README.md:96]:
```suggestion
### Clicking on links in mails
```

File: guides/testing-with-test-dispatch.md [hunk 6]
```diff
@@ -162,3 +162,34 @@ As expected here a post request is done to `/posts/1/comments/2` for the second
...
+
+When the mail is not received it will raise an error. Specific emails can be
+targeted by adding the `:subject`, `:to` or `:from` to the second argument of
+receive mail in a map.
+
+```elixir
+receive_mail(conn, %{submit: "This exact message", to: "this_address@exmaple.com"})
+```
+
+
```
Comment (fatboypunk) [inline guides/testing-with-test-dispatch.md:195]:
```suggestion
```

File: test/receive_emeail_test.exs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+  describe "receive_mail" do
+    test "receive email with expected_fields: [subject, to, from]", %{conn: conn} do
+      assert conn
+             |> get("/posts/1")
+             |> click_link("post-123-send-as-mail")
+             |> receive_mail(%{subject: "Post 1", to: "other@example.com", from: "me@app.com"})
+    end
+
+    test "receive email with expected_fields:: [subject, to]", %{conn: conn} do
```
Comment (EasterPeanut) [inline test/receive_emeail_test.exs:12]:
```suggestion
    test "receive email with expected_fields: [subject, to]", %{conn: conn} do
```

File: test/receive_emeail_test.exs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      assert conn
+             |> get("/posts/1")
+             |> click_link("post-123-send-as-mail")
+             |> receive_mail(%{subject: "Post 1", to: "other@example.com"})
+    end
+
+    test "receive email with expected_fields:: [subject, from]", %{conn: conn} do
```
Comment (EasterPeanut) [inline test/receive_emeail_test.exs:19]:
```suggestion
    test "receive email with expected_fields: [subject, from]", %{conn: conn} do
```

File: test/receive_emeail_test.exs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      assert conn
+             |> get("/posts/1")
+             |> click_link("post-123-send-as-mail")
+             |> receive_mail(%{subject: "Post 1", from: "me@app.com"})
+    end
+
+    test "receive email with expected_fields:: [to, from]", %{conn: conn} do
```
Comment (EasterPeanut) [inline test/receive_emeail_test.exs:26]:
```suggestion
    test "receive email with expected_fields: [to, from]", %{conn: conn} do
```

File: test/receive_emeail_test.exs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+      assert conn
+             |> get("/posts/1")
+             |> click_link("post-123-send-as-mail")
+             |> receive_mail(%{to: "other@example.com", from: "me@app.com"})
+    end
+
+    test "receive email with expected_fields:: [from]", %{conn: conn} do
```
Comment (EasterPeanut) [inline test/receive_emeail_test.exs:33]:
```suggestion
    test "receive email with expected_fields: [from]", %{conn: conn} do
```

File: test/receive_emeail_test.exs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+      assert conn
+             |> get("/posts/1")
+             |> click_link("post-123-send-as-mail")
+             |> receive_mail(%{from: "me@app.com"})
+    end
+
+    test "receive email with expected_fields:: [to]", %{conn: conn} do
```
Comment (EasterPeanut) [inline test/receive_emeail_test.exs:40]:
```suggestion
    test "receive email with expected_fields: [to]", %{conn: conn} do
```

File: test/receive_emeail_test.exs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+      assert conn
+             |> get("/posts/1")
+             |> click_link("post-123-send-as-mail")
+             |> receive_mail(%{to: "other@example.com"})
+    end
+
+    test "receive email with expected_fields:: [subject]", %{conn: conn} do
```
Comment (EasterPeanut) [inline test/receive_emeail_test.exs:47]:
```suggestion
    test "receive email with expected_fields: [subject]", %{conn: conn} do
```

**Review by EasterPeanut — CHANGES_REQUESTED** [review]

**Comment by EasterPeanut** [comment]
just grammar things

File: guides/testing-with-test-dispatch.md [hunk 13]
```diff
@@ -162,3 +162,32 @@ As expected here a post request is done to `/posts/1/comments/2` for the second
...
 If there is no `data-method` set `click_link/3` will do a `get` request by
 default.
+
+## Clicking on links in mails
+
+During the tests emails might be send that we want to integrate in our flow. For
+that there is `receive_mail/2`. It expects the conn as the first argument and
+the found email will be added to the conn as the `resp_body`. By using the conn
+for this, this can now be combined with the `click_link/4` function and "click" on
+the link in an email.
```
Comment (fatboypunk) [inline guides/testing-with-test-dispatch.md:172]:
```suggestion
the found email will be added to the conn as the `resp_body`. Using the conn
combined with the `click_link/4` function you can simulate "clicking" on the
link in an email.
```

**Review by EasterPeanut — APPROVED** [review]

**Pull request merged**